Fix part of #6240: Add E2E Tests for Profile Page Personal Deta…

U8NWXD opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Add protractor end-to-end tests for a user's personal details on their profile. These tests cover:

* Profile photo (not customized)
* Bio
* Subject Interests

in the following scenarios

* Profile has been customized, and the user is viewing their own page
* Profile has been customized, and the user is viewing someone else's profile
* Profile has not been customized, and the user is viewing their own page
* Profile has not been customized, and the user is viewing someone else's profile

Fixes part of #6240. 

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

U8NWXD committed:
Protractor: Default Profile of Current User

Create protractor end-to-end tests to check that when a logged-in
user has not customized their profile, their profile page shows:

* A profile photo
* Placeholder bio text
* Placeholder interests text

U8NWXD committed:
Protractor: Default Profile of Other User

Create protractor end-to-end tests to check that when not logged in,
the profile page of a user who has not customized their profile shows:

* No profile photo
* Placeholder bio text
* Placeholder interests text

U8NWXD committed:
Protractor: Customized Profile of Current User

Create protractor end-to-end tests to check that when logged in,
the profile page of a user who has customized their profile shows:

* Their profile photo
* Their customized bio text
* Their customized interests

U8NWXD committed:
Protractor: Customized Profile of Other User

Create protractor end-to-end tests to check that when not logged in,
the profile page of a user who has customized their profile shows:

* No profile photo
* Their customized bio text
* Their customized interests

U8NWXD committed:
Merge remote-tracking branch 'upstream/develop' into protractor-profile-features

U8NWXD committed:
Merge remote-tracking branch 'upstream/develop' into protractor-profile-features

seanlip commented:
@U8NWXD Could you please explain why this is a "Draft" PR? It looks like it's in a state that's ready for review, but we won't be assigning reviewers until it comes out of draft status...

U8NWXD commented:
@seanlip I used `protractor-*` for the HTML classes I created instead of `protractor-test-*`. I'll mark the PR as ready for review (out of draft status) once I fix that (hopefully within the next 12 hours)

U8NWXD committed:
Fix HTML Class Label Naming

Change the HTML classes used in the profileFeatures protractor
e2e tests to follow the `protractor-test-*` naming scheme specified in
the
[code owners wiki](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers#apb7)

U8NWXD committed:
Merge remote-tracking branch 'upstream/develop' into protractor-profile-features

U8NWXD commented:
I can't add labels to this PR, I'm guessing because I'm a new contributor and not a collaborator. Could someone add the appropriate labels?

seanlip commented:
Done. Thanks for flagging, @U8NWXD!

seanlip commented on core/tests/protractor_desktop/profileFeatures.js:
```diff
@@ -15,3 +15,178 @@
+    profilePage.get(TEST_USERNAME);
+  });
+
+  it('displays profile photo', function() {
```
For e2e tests, try to batch the "it"s a little bit, since otherwise the beforeEach and afterEach run each time. We normally keep them separate for unit tests, but since the setup/teardown takes an order of magnitude longer for e2e tests, we sometimes batch the tests together if it doesn't cause problems.

seanlip replied:
So e.g. I think these four "it"s could be condensed into one.

U8NWXD replied:
Good to know. Thanks for the explanation! Resolved in ddce9cd.

U8NWXD committed:
Merge remote-tracking branch 'upstream/develop' into protractor-profile-features

U8NWXD committed:
Merge Test Assertions Into Single `it`

Addresses review comments from @seanlip. Combining the test assertions
improves performance by avoiding unnecessary setup and teardown steps.

nithusha21 commented on core/templates/dev/head/pages/profile-page/profile-page.directive.html:
```diff
@@ -12,9 +12,12 @@
           </a>
-          <img ng-src="<[$ctrl.profilePictureDataUrl]>" class="oppia-profile-picture-fullsize" alt="">
+          <img ng-src="<[$ctrl.profilePictureDataUrl]>"
+               class="oppia-profile-picture-fullsize protractor-test-profile-curr-user-photo"
```
I think it is ok to spell "current" out completely.

U8NWXD replied:
Fixed in a564fbf

nithusha21 commented on core/templates/dev/head/pages/profile-page/profile-page.directive.html:
```diff
@@ -56,10 +59,13 @@ <h2 class="oppia-profile-username-small-screen">
         <span ng-repeat="interest in $ctrl.subjectInterests track by $index">
-          <span class="oppia-profile-subject-interest"><[interest]></span>
+          <span
+                class="oppia-profile-subject-interest protractor-test-profile-interest"
```
why break this to the newline?

U8NWXD replied:
Fixed in a564fbf

U8NWXD commented:
@nithusha21 PTAL

nithusha21 commented on core/tests/protractor_desktop/profileFeatures.js:
```diff
@@ -15,3 +15,142 @@
+var ProfilePage = require('../protractor_utils/ProfilePage.js');
+var PreferencesPage = require('../protractor_utils/PreferencesPage.js');
+
+describe('Un-customized profile page for current, logged-in user', function() {
```
I think we can combine the various describe blocks here into one big describe with several it() blocks right? The setup mostly seems common across all the suites here.

U8NWXD replied:
I think we should still have two `describe` blocks: one for unmodified profiles and one for modified profiles. My reasoning is that for the modified profiles, I include code to set the user's bio and interests in the `beforeAll` function, which shouldn't happen for the unmodified profiles. I do this in a564fbf.

nithusha21 reviewed (commented):
Hi @U8NWXD really sorry for the delayed response, was caught up with exams! Overall this looks great! Just a few comments.

U8NWXD committed:
Clean Up Profile Features e2e Tests

Improve code style and reduce redundancies in profile features tests.

Fixes part of #6240. Addresses review comments from @nithusha21.

U8NWXD commented:
@nithusha21 PTAL

seanlip commented:
@kevinlee12 @nithusha21 @vojtechjelinek PTAL as codeowners.

(@U8NWXD -- just a note, please always make sure that a PR has an assignee, so that it's clear who needs to take the next action. Thanks!)

nithusha21 reviewed (approved):
LGTM thanks!

kevinlee12 reviewed (approved):
lgtm!

vojtechjelinek reviewed (approved):
LGTM!
